### PR TITLE
minimal RStan package vignette

### LINKED ISSUE #9: Add vignette that shows steps required to create a functioning package

### PR DESCRIPTION
**Pull request by sieste**
Hi there,

You did a great job on the developer guidelines vignette. But it still took me a while to figure out how to actually develop a working package that depends on RStan. To help future users, I wrote it down as an additional vignette. Maybe you find it useful.

This pull request fixes #9.

Cheers,
Stefan

### CODE REVIEW

**Comment by jgabry** [comment]
Thanks! Haven’t had a chance to take a look in detail yet but we’ve been
meaning to add another vignette like this so it’s great to see this PR. One
of us will go over this sometime soon and review the PR.

**Comment by jgabry** [comment]
@bgoodri Do any of the changes we need to make to `rstan_package_skeleton` for the next release conflict with the content in the vignette in this PR?

File: vignettes/minimal-rstan-package.Rmd [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+## Creating a minimal package that uses Stan, step-by-step
+
+This vignette is a step-by-step guide to create a functioning R package that uses Stan.
+It is complementary to the [rstantools developer guidelines vignette](developer-guidelines.html) which you should read first.
+
+
+To start off, we use `rstan_package_skeleton` to initialise a bare-bones package directory.
+The name of our package will be `rstanlm`; it will fit a simple linear regression model using Stan.
+Under the hood, the function `rstan_package_skeleton` calls `utils::package.skeleton()` which won't allow creating a completely empty package directory, so we initialise it with a dummy variable which we will delete later on:
+
```
Comment (jgabry) [inline vignettes/minimal-rstan-package.Rmd]: I think you can call `rstan_package_skeleton("rstanlm")` without the dummy environment list. Does that result in an error for you?

File: vignettes/minimal-rstan-package.Rmd [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+The newly created package directory is called `rstanlm` and we change to it to make a few modifications:
+
+```{r}
+setwd('rstanlm')
+```
+
+We remove the following files, which are either unneeded, or will be automatically recreated later on:
+
+```{r}
+file.remove('Read-and-delete-me')
```
Comment (jgabry) [inline vignettes/minimal-rstan-package.Rmd]: I think we should recommend at least glancing at the 'Read-and-delete-me' file before deleting it, even if the rest of this vignette essentially covers the content of the file

File: vignettes/minimal-rstan-package.Rmd [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+file.remove('data/dummy.rda')
+file.remove('man/dummy.Rd')
```
Comment (jgabry) [inline vignettes/minimal-rstan-package.Rmd]: should't be necessary if not including the dummy environment in the call to `rstan_package_skeleton`

File: vignettes/minimal-rstan-package.Rmd [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+file.remove('man/rstanlm-package.Rd')
+```
+
+Some of the magic to make Stan models available as part of your package happens in the file `cleanup` (or `cleanup.win` on Windows).
+The `cleanup` script is executed during the build process, but for this to work its file mode has to be changed to executable:
+
+```{r}
+Sys.chmod('cleanup', mode='755')
+Sys.chmod('cleanup.win', mode='755')
+```
```
Comment (jgabry) [inline vignettes/minimal-rstan-package.Rmd]: @bgoodri Is there no alternative to manually changing the file modes like this? We don't mention the need to do this anywhere in the read-and-delete-me file

File: vignettes/minimal-rstan-package.Rmd [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  y ~ normal(intercept + beta * x, sigma);
+}
+```
+
+The `exec` subdirectory can contain further Stan models. 
+During installation, all Stan models will be compiled and saved in the list `stanmodels` that can then be used by R function in the package.
+The rule is that the Stan model compiled from the model code in `exec/foo.stan` is stored as list element `stanmodels$foo`.
+
+We next create the file `R/lm_stan.R` where we define the function `lm_stan` in which our compiled Stan model is being used.
+A comment block in roxygen2 syntax ensures that the function has a help file and that it is added to the NAMESPACE:
```
Comment (jgabry) [inline vignettes/minimal-rstan-package.Rmd]: Can you make 'roxygen2' a hyperlink: `[roxygen2](https://CRAN.R-project.org/package=roxygen2)`

File: vignettes/minimal-rstan-package.Rmd [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+We create the top-level package file `R/rstanlm-package.R` where we define all dependencies, imports, and briefly decribe the functionality of our package:
+
+```{r}
+#' RStan Linear Regression
+#'
+#' @docType package
+#' @name rstanlm-package
+#' @aliases rstanlm
+#' @useDynLib rstanlm, .registration = TRUE
```
Comment (jgabry) [inline vignettes/minimal-rstan-package.Rmd]: Maybe emphasize not to forget this @useDynLib line!

File: vignettes/minimal-rstan-package.Rmd [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+Date: 2017-11-30
+Author: Joe Developer [aut, cre]
+Maintainer: Joe Developer <joe@developer.net>
+Description: This is a minimal working example of a package that uses RStan. The 
+             package enables you to fit a linear regression model with RStan. 
+License: GPL (>=3)
+Depends: R (>= 3.0.2), Rcpp (>= 0.12.14), methods
+Imports: rstan (>= 2.16.2), rstantools (>= 1.3.0)
+LinkingTo: StanHeaders (>= 2.16.0.1), rstan (>= 2.16.2), BH (>= 1.65.0.1), 
+           Rcpp (>= 0.12.14), RcppEigen (>= 0.3.3.3.0)
```
Comment (jgabry) [inline vignettes/minimal-rstan-package.Rmd]: Can you update this example DESCRIPTION file so that the date and package versions are current? And

**Comment by jgabry** [comment]
Actually, we're making some improvements to `rstan_package_skeleton` that will necessitate changing a little bit of the content in the vignette. I'm going to merge this and then I can make the necessary edits based on the changes to `rstan_package_skeleton`.

**Comment by jgabry** [comment]
@sieste so ignore all the comments on the PR. We’ll take care of the edits, but I will add you in the package DESCRIPTION as a contributor.

**Comment by sieste** [comment]
Excellent, thanks for accepting the PR and including me as contributor. This is highly appreciated! 

I left a few comments on the commit 291b0d7641a6f127a1a336983d39e570b5e92d23 and opened two issues (#24 and #25).

**Pull request merged**